File: DataSheets/FlowPathwayDiagramDataSheet.vb
'************************************************************************************
' STSimStockFlow: A .NET library for simulating stocks and flows
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core
Imports System.Globalization
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class FlowPathwayDiagramDataSheet
    Inherits DataSheet

    Public Const ERROR_INVALID_CELL_ADDRESS As String =
        "The value must be a valid cell address (a valid cell address is a letter from 'A' to 'Z' followed by a number from 1 to 255.  Example: 'A25')"

    Public Overrides Sub Validate(ByVal proposedValue As Object, columnName As String)

        MyBase.Validate(proposedValue, columnName)

        If (columnName = LOCATION_COLUMN_NAME) Then

            If (Not IsValidLocation(proposedValue)) Then
                Throw New DataException(ERROR_INVALID_CELL_ADDRESS)
            End If

        End If

    End Sub

    Public Overrides Function GetDeleteRowsConfirmation() As String

        Dim m As String = MyBase.GetDeleteRowsConfirmation()

        If (String.IsNullOrWhiteSpace(m)) Then
            m = "Associated flows will also be deleted.  Continue?"
        End If

        Return m

    End Function

    Protected Overrides Sub OnRowsDeleted(sender As Object, e As SyncroSim.Core.DataSheetRowEventArgs)

        Dim DeletedRows As Boolean = False
        Dim RemainingStockTypes As Dictionary(Of Integer, Boolean) = CreateRecordLookup(Me, STOCK_TYPE_ID_COLUMN_NAME)
        Dim FlowPathwaySheet As DataSheet = Me.GetDataSheet(DATASHEET_FLOW_PATHWAY_NAME)
        Dim FlowPathwayData As DataTable = FlowPathwaySheet.GetData()

        For i As Integer = FlowPathwayData.Rows.Count - 1 To 0 Step -1

            Dim dr As DataRow = FlowPathwayData.Rows(i)

            If (dr.RowState = DataRowState.Deleted) Then
                Continue For
            End If

            Dim FromStockTypeId As Integer = CInt(dr(FROM_STOCK_TYPE_ID_COLUMN_NAME))
            Dim ToStockTypeId As Integer = CInt(dr(TO_STOCK_TYPE_ID_COLUMN_NAME))

            If ((Not RemainingStockTypes.ContainsKey(FromStockTypeId)) Or
                (Not RemainingStockTypes.ContainsKey(ToStockTypeId))) Then

                DataTableUtilities.DeleteTableRow(FlowPathwayData, dr)
                DeletedRows = True

            End If

        Next

        If (DeletedRows) Then

            FlowPathwaySheet.Changes.Add(
                New ChangeRecord(Me, "Diagram data deleted rows"))

        End If

        MyBase.OnRowsDeleted(sender, e)

    End Sub

    Private Shared Function IsValidLocation(ByVal proposedLocation As Object) As Boolean

        If (proposedLocation Is Nothing) Then
            Return False
        End If

        Dim Location As String = CStr(proposedLocation)

        If (String.IsNullOrEmpty(Location)) Then
            Return False
        End If

        Dim LocUpper As String = Location.ToUpper(CultureInfo.CurrentCulture).Trim

        If (String.IsNullOrEmpty(LocUpper)) Then
            Return False
        End If

        If (LocUpper.Length < 2) Then
            Return False
        End If

        Dim CharPart As String = Mid(LocUpper, 1, 1)
        Dim NumPart As String = Mid(LocUpper, 2, LocUpper.Length - 1)

        If (String.IsNullOrEmpty(CharPart) Or String.IsNullOrEmpty(NumPart)) Then
            Return False
        End If

        If (Asc(CharPart) < Asc("A") Or Asc(CharPart) > Asc("Z")) Then
            Return False
        End If

        Dim n As Integer = 0
        If (Not Integer.TryParse(NumPart, n)) Then
            Return False
        End If

        If (n <= 0 Or n > 256) Then
            Return False
        End If

        Return True

    End Function

End Class

File: DataSheets/FlowPathwayDiagramDataSheet.cs
//*********************************************************************************************
// STSimStockFlow: A SyncroSim Module for the ST-Sim Stocks and Flows Add-In.
//
// Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
//
//*********************************************************************************************

using System;
using System.Data;
using System.Reflection;
using SyncroSim.Core;
using System.Globalization;
using System.Collections.Generic;

namespace SyncroSim.STSimStockFlow
{
	[ObfuscationAttribute(Exclude=true, ApplyToMembers=false)]
	internal class FlowPathwayDiagramDataSheet : DataSheet
	{
		public const string ERROR_INVALID_CELL_ADDRESS = "The value must be a valid cell address (a valid cell address is a letter from 'A' to 'Z' followed by a number from 1 to 255.  Example: 'A25')";

		public override void Validate(object proposedValue, string columnName)
		{
			base.Validate(proposedValue, columnName);

			if (columnName == Constants.LOCATION_COLUMN_NAME)
			{
				if (!IsValidLocation(proposedValue))
				{
					throw new DataException(ERROR_INVALID_CELL_ADDRESS);
				}
			}
		}

		public override string GetDeleteRowsConfirmation()
		{
			string m = base.GetDeleteRowsConfirmation();

			if (string.IsNullOrWhiteSpace(m))
			{
				m = "Associated flows will also be deleted.  Continue?";
			}

			return m;
		}

		protected override void OnRowsDeleted(object sender, SyncroSim.Core.DataSheetRowEventArgs e)
		{
			bool DeletedRows = false;
			Dictionary<int, bool> RemainingStockTypes = LookupKeyUtilities.CreateRecordLookup(this, Constants.STOCK_TYPE_ID_COLUMN_NAME);
			DataSheet FlowPathwaySheet = this.GetDataSheet(Constants.DATASHEET_FLOW_PATHWAY_NAME);
			DataTable FlowPathwayData = FlowPathwaySheet.GetData();

			for (int i = FlowPathwayData.Rows.Count - 1; i >= 0; i--)
			{
				DataRow dr = FlowPathwayData.Rows[i];

				if (dr.RowState == DataRowState.Deleted)
				{
					continue;
				}

				int FromStockTypeId = Convert.ToInt32(dr[Constants.FROM_STOCK_TYPE_ID_COLUMN_NAME]);
				int ToStockTypeId = Convert.ToInt32(dr[Constants.TO_STOCK_TYPE_ID_COLUMN_NAME]);

				if ((!RemainingStockTypes.ContainsKey(FromStockTypeId)) | (!RemainingStockTypes.ContainsKey(ToStockTypeId)))
				{
					DataTableUtilities.DeleteTableRow(FlowPathwayData, dr);
					DeletedRows = true;
				}
			}

			if (DeletedRows)
			{
				FlowPathwaySheet.Changes.Add(new ChangeRecord(this, "Diagram data deleted rows"));
			}

			base.OnRowsDeleted(sender, e);
		}

		private static bool IsValidLocation(object proposedLocation)
		{
			if (proposedLocation == null)
			{
				return false;
			}

			string Location = Convert.ToString(proposedLocation);

			if (string.IsNullOrEmpty(Location))
			{
				return false;
			}

			string LocUpper = Location.ToUpper(CultureInfo.InvariantCulture).Trim();

			if (string.IsNullOrEmpty(LocUpper))
			{
				return false;
			}

			if (LocUpper.Length < 2)
			{
				return false;
			}

			string CharPart = LocUpper.Substring(0, 1);
			string NumPart = LocUpper.Substring(1, LocUpper.Length - 1);

			if (string.IsNullOrEmpty(CharPart) | string.IsNullOrEmpty(NumPart))
			{
				return false;
			}

			if (CharPart[0] < 'A' || CharPart[0] > 'Z')
			{
				return false;
			}

			int n = 0;
			if (!int.TryParse(NumPart, out n))
			{
				return false;
			}

			if (n <= 0 || n > 256)
			{
				return false;
			}

			return true;
		}
	}
}